Assets/HoloToolkit/UX/Scripts/Lines/Rectangle.cs

```diff
@@ -0,0 +1,196 @@
+
+            return LineUtils.InterpolateVectorArray(points, normalizedDistance);
+        }
+
+        protected override void SetPointInternal(int pointIndex, Vector3 point)
+        {
+            if (points == null || points.Length != 8)
+                points = new Vector3[8];
+
+            if (pointIndex <= 7 && pointIndex >= 0)
+                points[pointIndex] = point;
+        }
+
+        protected override Vector3 GetPointInternal(int pointIndex)
+        {
+            if (points == null || points.Length != 8)
+                points = new Vector3[8];
+
+            if (pointIndex <= 7 && pointIndex >= 0)
+                    return points[pointIndex];
+
+            return Vector3.zero;
+        }
+
+        protected override float GetUnclampedWorldLengthInternal()
+        {
+            BuildPoints();
+
+            Vector3 last = points[0];
+            float distance = 0f;
+            for (int i = 1; i < points.Length; i++)
+            {
+                distance += Vector3.Distance(last, points[i]);
+                last = points[i];
+            }
+            return distance;
+        }
+
+        protected override Vector3 GetUpVectorInternal(float normalizedLength)
+        {
+            // Rectangle 'up' vector always points out from center
+            return (GetPoint(normalizedLength) - transform.position).normalized;
+        }
+
+        private void BuildPoints()
+        {
+            Vector3 offset = Vector3.forward * ZOffset;
+            Vector3 top = (Vector3.up * YSize * 0.5f);
+            Vector3 bot = (Vector3.down * YSize * 0.5f);
+            Vector3 left = (Vector3.left * XSize * 0.5f);
+            Vector3 right = (Vector3.right * XSize * 0.5f);
+
+            SetPointInternal(0, top + left + offset);
+            SetPointInternal(1, top + offset);
+            SetPointInternal(2, top + right + offset);
+            SetPointInternal(3, right + offset);
+            SetPointInternal(4, bot + right + offset);
+            SetPointInternal(5, bot + offset);
+            SetPointInternal(6, bot + left + offset);
+            SetPointInternal(7, left + offset);
+        }
+
+#if UNITY_EDITOR
+        [UnityEditor.CustomEditor(typeof(Rectangle))]
+        public class CustomEditor : LineBaseEditor {
+            // Use FromSource step mode for rectangles since interpolated looks weird
+            protected override StepModeEnum EditorStepMode { get { return StepModeEnum.FromSource; } }            
+        }
+
+        protected override void OnDrawGizmos()
+        {
+            // Show gizmos if this object is not selected
+            // (SceneGUI will display it otherwise)
+
+            if (Application.isPlaying)
+                return;
+
+            if (UnityEditor.Selection.activeGameObject == this.gameObject)
+                return;
+
+            // Only draw a gizmo if we don't have a line renderer
+            LineRendererBase lr = gameObject.GetComponent<LineRendererBase>();
```

Comment:
we should cache this instead of getting the component each update.

Reply:
Normally I'd agree but I don't want to add a new field to the class for the sake of a Gizmos function. Given that Lines can have multiple LineRenderers I worry that its presence would be confusing for devs.

Since we only do the GetComponent call while the editor isn't playing it won't be a performance issue.